### PR TITLE
v1: Add upgrade occurrence and note type.

### PR DESCRIPTION
**Pull request by clrprod**
This type represents a available software upgrade for a given resource
url. It is complementary to the package and vulnerability types.

### CODE REVIEW

File: proto/v1/upgrade.proto [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+package grafeas.v1;
+
+option go_package = "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1";
+option objc_class_prefix = "GRA";
+
+import "proto/v1/package.proto";
+
+message UpgradeNote {
```
Comment (vtsao) [inline proto/v1/upgrade.proto]: Can you add comments describing each message?

File: proto/v1/upgrade.proto [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  // Required - The package this Upgrade is for.
+  string package = 1;
+  // Required - The version of the package in machine + human readable form
```
Comment (vtsao) [inline proto/v1/upgrade.proto]: nit: periods at the end of each comment

**Comment by vtsao** [comment]
We're waiting for @aysylu to LGTM this as well.

File: proto/v1/upgrade.proto [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+option go_package = "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1";
+option objc_class_prefix = "GRA";
+
+import "proto/v1/package.proto";
+
+// An upgrade note represents a potential upgrade of a package to a given version.
+// In general there should be an upgrade note for every single package version combination.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: nit: comma after "In general"

File: proto/v1/upgrade.proto [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+package grafeas.v1;
+
+option go_package = "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1";
+option objc_class_prefix = "GRA";
+
+import "proto/v1/package.proto";
+
+// An upgrade note represents a potential upgrade of a package to a given version.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: nit: please capitalize "Upgrade Note" for consistency

File: proto/v1/upgrade.proto (hunk 3 above)
Comment (aysylu) [inline proto/v1/upgrade.proto]: This doesn't read clearly. Could you please clarify?

File: proto/v1/upgrade.proto [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// In general there should be an upgrade note for every single package version combination.
+message UpgradeNote {
+  // Required - The package this Upgrade is for.
+  string package = 1;
+  // Required - The version of the package in machine + human readable form,
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: s/,/.

File: proto/v1/upgrade.proto [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  grafeas.v1.Version version = 2;
+  // Per operating system metadata.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: This doesn't read clearly, perhaps rephrase as "for every operating system metadata"?

File: proto/v1/upgrade.proto [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  repeated UpgradeDistribution distributions = 3;
+}
+
+// The Upgrade Distribution represents per cpe metadata for a given upgrade. Some distributions
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: Reading this, "per cpe metadata for a given upgrade" doesn't make sense. Could you please rephrase? Also, s/upgrade/Upgrade for consistency.

File: proto/v1/upgrade.proto [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+// have additional metadata around updates, classifying them into various string categories and
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: The type that represents categories shouldn't be part of the high-level comment for something in the API.

File: proto/v1/upgrade.proto [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+// severities.
+message UpgradeDistribution {
+  // Required - The specific operating system this metadata applies to.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: please add: ", e.g. <any os example>"

File: proto/v1/upgrade.proto [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  string cpe_uri = 1;
+  // Optional - The operating system classification of this Upgraded, as specified by the upstream operating system upgrade feed.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: No need to mention a field is optional in proto3, it is by default, unless we comment that it's required, to be consistent across the codebase.

Comment (aysylu) [inline proto/v1/upgrade.proto]: s/Upgraded/Upgrade

Comment (aysylu) [reply proto/v1/upgrade.proto]: ditto for below

File: proto/v1/upgrade.proto [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  string classification = 2;
+  // Optional - The severity as specified by the upstream operating system.
+  string severity = 3;
+  // Optional - The cve tied to this Upgrade.
+  repeated string cve = 4;
+}
+
+// An upgrade occurrence represents that a specific resourceurl could install a specific upgrade
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: s/resourceurl/resource_url, for consistency

Comment (aysylu) [inline proto/v1/upgrade.proto]: add period at the end of the sentence, for consistency

File: proto/v1/upgrade.proto [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+// This presence is supplied via local sources (i.e. it is present in the mirror and the running system has noticed it's availability).
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: s/it's/its

File: proto/v1/upgrade.proto [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+message UpgradeOccurrence {
+  // Required - The package this Upgrade is for.
+  string package = 1;
+  // Required - The version of the package in a machine + human readable form
+  grafeas.v1.Version parsed_version = 3;
+  // Per operating system metadata
+  UpgradeDistribution distribution = 4;
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: Ditto here for the comment. It'd be good to include why we're duplicating this field in the Occurrence.

File: proto/v1/upgrade.proto (hunk 1 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 2 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 3 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 4 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done

File: proto/v1/upgrade.proto (hunk 3 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Rewrote to use an example and hopefully be more clear.

File: proto/v1/upgrade.proto (hunk 5 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done

File: proto/v1/upgrade.proto (hunk 6 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 7 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Reworded

File: proto/v1/upgrade.proto (hunk 8 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Removed string

File: proto/v1/upgrade.proto (hunk 10 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 9 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Added a link to match vulnerability.proto.

File: proto/v1/upgrade.proto (hunk 10 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done

File: proto/v1/upgrade.proto (hunk 11 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done

Comment (clrprod) [reply proto/v1/upgrade.proto]: Done.

File: proto/v1/upgrade.proto (hunk 12 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Done

File: proto/v1/upgrade.proto (hunk 13 above)
Comment (clrprod) [reply proto/v1/upgrade.proto]: Updated comment.

File: proto/v1/common.proto [hunk 14]
```diff
@@ -39,6 +39,8 @@ enum NoteKind {
   DISCOVERY = 6;
   // This represents a logical "role" that can attest to artifacts.
   ATTESTATION = 7;
+  // This represents a available package upgrade.
```
Comment (aysylu) [inline proto/v1/common.proto]: nit: a/an

File: proto/v1/upgrade.proto [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+option go_package = "github.com/grafeas/grafeas/proto/v1/grafeas_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1";
+option objc_class_prefix = "GRA";
+
+import "proto/v1/package.proto";
+
+// An Upgrade Note represents a potential upgrade of a package to a given version.
+// For each package version combination (i.e. bash 4.0, bash 4.1, bash 4.1.2),
+// there will be a upgrade note.
```
Comment (aysylu) [inline proto/v1/upgrade.proto]: nit: s/a/an and capitalization for Upgrade Note, for consistency

**Pull request merged**